test/unit/BPool.t.sol

```diff
@@ -0,0 +1,573 @@
+
+  function test_Revert_NotBoundTokenOut() private view {}
+
+  function test_Returns_SpotPrice() private view {}
+
+  function test_Revert_Reentrancy() private view {}
+}
+
+contract BPool_Unit_JoinPool is Base {
+  struct FuzzScenario {
+    uint256 poolAmountOut;
+    uint256 initPoolSupply;
+    uint256[TOKENS_AMOUNT] balance;
+  }
+
+  function test_HappyPath(FuzzScenario memory _fuzz) public happyPath(_fuzz) {
+    uint256[] memory maxAmountsIn = new uint256[](tokens.length);
+    for (uint256 i = 0; i < tokens.length; i++) {
+      maxAmountsIn[i] = type(uint256).max;
+    } // Using max possible amounts
+
+    bPool.joinPool(_fuzz.poolAmountOut, maxAmountsIn);
+  }
+
+  function test_Revert_NotFinalized() private view {}
+
+  function test_Revert_MathApprox() private view {}
+
+  function test_Revert_TokenArrayMathApprox() private view {}
+
+  function test_Revert_TokenArrayLimitIn() private view {}
+
+  function test_Revert_Reentrancy() private view {}
+
+  function test_Set_TokenArrayBalance() private view {}
+
+  function test_Emit_TokenArrayLogJoin() private view {}
+
+  function test_Pull_TokenArrayTokenAmountIn() private view {}
+
+  function test_Mint_PoolShare() private view {}
+
+  function test_Push_PoolShare() private view {}
+
+  function test_Emit_LogCall() private view {}
+
+  modifier happyPath(FuzzScenario memory _fuzz) {
+    _assumeHappyPath(_fuzz);
+    _setValues(_fuzz);
+    _;
+  }
+
+  function _setValues(FuzzScenario memory _fuzz) internal {
+    // Create mocks
+    for (uint256 i = 0; i < tokens.length; i++) {
+      vm.mockCall(tokens[i], abi.encodeWithSelector(IERC20(tokens[i]).transfer.selector), abi.encode(true));
+      vm.mockCall(tokens[i], abi.encodeWithSelector(IERC20(tokens[i]).transferFrom.selector), abi.encode(true));
+    }
+
+    // Set tokens
+    bytes memory _arraySlot = abi.encode(9);
+    bytes32 _hashArraySlot = keccak256(_arraySlot);
+    vm.store(address(bPool), bytes32(_arraySlot), bytes32(tokens.length)); // write length
+    for (uint256 i = 0; i < tokens.length; i++) {
+      vm.store(address(bPool), bytes32(uint256(_hashArraySlot) + i), bytes32(abi.encode(tokens[i]))); // write token
+    }
+
+    // Set balances
+    for (uint256 i = 0; i < tokens.length; i++) {
+      bytes32 _slot = keccak256(abi.encode(tokens[i], 10)); // mapping is found at slot 10
+      vm.store(address(bPool), bytes32(uint256(_slot) + 0), bytes32(abi.encode(1))); // bound
+      vm.store(address(bPool), bytes32(uint256(_slot) + 3), bytes32(abi.encode(_fuzz.balance[i]))); // balance
+    }
+
+    // Set public swap
+    vm.store(
+      address(bPool),
+      bytes32(uint256(6)),
+      bytes32(uint256(0x0000000000000000000000010000000000000000000000000000000000000000))
+    );
+    // Set finalize
+    vm.store(address(bPool), bytes32(uint256(8)), bytes32(uint256(1)));
+    // Set totalSupply
+    vm.store(address(bPool), bytes32(uint256(2)), bytes32(_fuzz.initPoolSupply));
+  }
```

Comment:
this methods may come handy for other scenarios and other happy paths, lets internalize in a higher hierarchy contract so that we can reuse them later :pray: